compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/loops/HeaderProcessor.kt

```diff
@@ -265,30 +262,17 @@ internal class HeaderProcessor(
         // Get the iterable expression, e.g., `someIterable` in the following loop variable declaration:
         //
         //   val it = someIterable.iterator()
-        //
-        // If the `iterator` method is an extension method, make sure that we are calling a known extension
-        // method in the library such as `kotlin.text.StringsKt.iterator` with no value arguments. Other
-        // extension methods could return user-defined iterators.
-        val iterable = (variable.initializer as? IrCall)?.let {
-            val extensionReceiver = it.extensionReceiver
+        val iteratorCall = variable.initializer as? IrCall
+        val iterable = iteratorCall?.run {
             if (extensionReceiver != null) {
-                val function = it.symbol.owner
-                if (it.valueArgumentsCount == 0
-                    && function.isTopLevel
-                    && function.getPackageFragment()?.fqName == FqName("kotlin.text")
-                    && function.name == OperatorNameConventions.ITERATOR
```

Comment:
Is this check is only actual for CharSequence?

Reply:
Yes, it's only for CharSequence. This is what we're trying to match: https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.text/iterator.html

Reply:
I'm not sure I understood your question correctly, but the reason I moved this into `CharSequenceIterationHandler` is because the handlers are only try to match the iterable (the subject of the for-loop). In the previous code, if there is a `CharSequence object` with a member `iterator()`, the value of `iterable` in this block of code would be the dispatch receiver (the `CharSequence`). Later on it would match in `CharSequenceIterationHandler` and attempt to lower the loop, which is wrong because that iterator's behavior is unknown.

By moving the matching of the `iterator()` call to the handlers, we are able to match both `iterator()` call and the iterable at the same time.

Reply:
I updated the test to cover more cases. I hope that helps clarify. Please let me know if I wasn't able to answer your question!

Reply:
Thank you for detailed explanation!